src/Providers/UserProviders.jsx

```diff
@@ -1,45 +1,99 @@
-import React, { Component, createContext } from "react";
-import { auth, generateUserDocument } from "../firebaseCredentials.js";
+import React, { useEffect, useState, createContext } from "react";
+import { auth } from "../firebaseCredentials.js";
 import { httpGet } from "../lib/dataAccess";
 
 const baseEndpoint = process.env.REACT_APP_ADMIN_BASE_ENDPOINT;
 
-export const UserContext = createContext({ user: null });
-class UserProvider extends Component {
-    state = {
+export const UserContext = createContext();
+
+const isTokenExpired = (user) => {
+    const MILLISECONDS_PER_DAY = 24 * 3600 * 1000;
+    const NOW = new Date().getTime();
+
+    // Force user to relog in every 24 hours
+    return Number(user.toJSON().lastLoginAt) < NOW - MILLISECONDS_PER_DAY;
+};
+
+const AUTH_STATES = {
+    LOADING: "LOADING",
+    NOT_AUTHENTICATED: "NOT_AUTHENTICATED",
+    AUTHENTICATED: "AUTHENTICATED",
+};
+
+export default function UserProvider({ children }) {
+    const [authState, setAuthState] = useState({
         user: null,
-        isAdmin: false,
-    };
+        state: AUTH_STATES.LOADING,
+    });
 
-    componentDidMount = async () => {
-        auth.onAuthStateChanged(async (userAuth) => {
-            if (!userAuth) {
-                this.setState({ user: null });
+    useEffect(() => {
+        async function setUser(user) {
+            //Set the state to loading by default
+            setAuthState({ state: AUTH_STATES.LOADING, user: null });
+
+            if (user == null) {
+                setAuthState({ state: AUTH_STATES.NOT_AUTHENTICATED, user });
+                return;
             }
-            if (userAuth && userAuth.emailVerified) {
-                const user = await generateUserDocument(userAuth);
-                this.setState({ user });
+
+            if (isTokenExpired(user)) {
+                auth.signOut();
+                setAuthState({
+                    state: AUTH_STATES.NOT_AUTHENTICATED,
+                    user: null,
+                });
+                return;
             }
 
-            // validate cookies are set appropriately on this user instance
-            try {
-                const response = await httpGet(baseEndpoint + "user_profile");
-                this.setState({ isAdmin: response.status === 200 });
-            } catch {
-                this.setState({ isAdmin: false });
+            const response = await httpGet(baseEndpoint + "user_profile");
```

Comment:
re: comment about race condition still being there

On login this useEffect gets run and this specific httpGet call errors out, then admin_login completes, and then a call is forced using reloadUser (see function on line 71).

One work around is to nest this call in a check that the cookies dcc_auth is set (I can do this with a cookies library) but I didn't want to add another dependency if not absolutely needed.

Reply:
Well, checking if the cookie is set doesn't solve it either. Users could have a cookie set, but it's invalid because it is expired. 
So here the problem is, you are parallelizing the call to admin_login and httpGET to user_profile. But in reality, you should be awaiting for the call to admin_login to end before calling user_profile, since there's no point doing this call if admin_login hasn't returned. So if you wanted to fix this, you would make changes to fix this specific ordering of calls.

Don't worry about this, as I said, let's get this merged in, been too long.